nexus/src/app/sagas/instance_create.rs

```diff
@@ -243,176 +217,145 @@ async fn sic_alloc_server(
         params::InstanceNetworkInterfaceAttachment::Default => {
-            Ok(vec![Uuid::new_v4()])
+            sic_create_default_primary_network_interface(&sagactx, nic_index)
+                .await
         }
         params::InstanceNetworkInterfaceAttachment::Create(
             ref create_params,
-        ) => {
-            if create_params.len() > MAX_NICS_PER_INSTANCE.try_into().unwrap() {
-                return Err(ActionError::action_failed(
-                    Error::invalid_request(
-                        format!(
-                            "Instances may not have more than {}
-                            network interfaces",
-                            MAX_NICS_PER_INSTANCE
-                        )
-                        .as_str(),
-                    ),
-                ));
+        ) => match create_params.get(nic_index) {
+            None => Ok(None),
+            Some(ref prs) => {
+                sic_create_custom_network_interface(&sagactx, prs).await
             }
-            let mut ids = Vec::with_capacity(create_params.len());
-            for _ in 0..create_params.len() {
-                ids.push(Uuid::new_v4());
-            }
-            Ok(ids)
-        }
+        },
     }
 }
 
-async fn sic_create_network_interfaces(
+/// Delete one network interface, by index.
+async fn sic_create_network_interface_undo(
     sagactx: ActionContext<SagaInstanceCreate>,
-) -> Result<(), ActionError> {
-    match sagactx.saga_params().create_params.network_interfaces {
-        params::InstanceNetworkInterfaceAttachment::None => Ok(()),
-        params::InstanceNetworkInterfaceAttachment::Default => {
-            sic_create_default_primary_network_interface(&sagactx).await
-        }
-        params::InstanceNetworkInterfaceAttachment::Create(
-            ref create_params,
-        ) => {
-            sic_create_custom_network_interfaces(&sagactx, &create_params).await
+    iface_index: usize,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let datastore = osagactx.datastore();
+    let saga_params = sagactx.saga_params();
+    let opctx =
+        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
+    let interface_id = sagactx.lookup::<Option<Uuid>>(
+        format!("network_interface{iface_index}").as_str(),
+    )?;
+    match interface_id {
+        None => Ok(()),
+        Some(id) => {
+            let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+            let (.., authz_instance) = LookupPath::new(&opctx, &datastore)
+                .instance_id(instance_id)
+                .lookup_for(authz::Action::Modify)
+                .await
+                .map_err(ActionError::action_failed)?;
+            let (.., authz_interface) = LookupPath::new(&opctx, &datastore)
+                .network_interface_id(id)
+                .lookup_for(authz::Action::Delete)
+                .await
+                .map_err(ActionError::action_failed)?;
```

Comment:
I *think* it's actually okay if we look up the interface and see "not found". Actions and undo actions are supposed to be idempotent, so let's suppose:

- Nexus boots, loads saga log, starts rolling back an instance_create saga
- We start the `sic_create_network_interface_undo` function. We find the interface, and delete it successfully.
- ... but before we record the "undo action" as complete in the saga log, we crash.

Now, nexus reboots. It'll retry the action, but since the interface was already deleted, it'll hit this "action_failed" case.